Yosemite/Yosemite/Stores/InAppPurchaseStore.swift

```diff
@@ -283,12 +283,24 @@ private extension InAppPurchaseStore {
                 return
             }
             for await result in Transaction.updates {
-                do {
-                    // Wait until the purchase finishes
-                    _ = await self.pauseTransactionListener.values.contains(false)
-                    try await self.handleCompletedTransaction(result)
-                } catch {
-                    self.logError("Error handling transaction update: \(error)")
+                switch result {
+                case .unverified:
+                    // Ignore unverified transactions.
+                    self.logError("Transaction unverified")
+                    break
+                case .verified(let transaction):
+                    do {
+                        _ = await self.pauseTransactionListener.values.contains(false)
+                        // For verified transactions, check whether a transaction has been handled already,
+                        // and mark it as finished if that's the case, or handle it otherwise
+                        _ = try await self.remote.retrieveHandledTransactionSiteID(for: transaction.id)
+                        await transaction.finish()
+
+                        self.logInfo("Marking transaction \(transaction.id) as finished")
+                    } catch {
+                        _ = await self.pauseTransactionListener.values.contains(false)
```

Comment:
Do we need to add this step (waiting for the transaction to be finished) twice in the `do` and `catch` blocks?

Reply:
We don't, modified here: https://github.com/woocommerce/woocommerce-ios/pull/10247/commits/edb5c19b292736def74a036ecdfdea040a8c4121